### PR TITLE
Parallelism for the exact solver

### PR DESCRIPTION
**Pull request by lquenti**

### CODE REVIEW

**Comment by lquenti** [comment]
See my mention at #21 sorry that I have to pollute this PR with `nearest_neighbour.rs` docs

**Comment by lquenti** [comment]
obviously verified that the MPI algo works on MPICH locally and on OpenMPI at the cluster!

**Comment by lquenti** [comment]
@johann-cm one thing I'd love to do is to finally make our mains more consistent by moving my mpi bootstrap code out of the calculating function and into the `lib.rs` and then use you `mpi::initialize()` checker to verify that MPI is correctly initialized.

But I'd love to touch this once we have the presentation done.

If it is fine with you, I'd put it into an issue and do it sometime next week when I have plenty time

(if you are okay with that, I'll create the issue myself no worries)

File: src/datastructures/mod.rs [hunk 1]
```diff
@@ -70,3 +73,8 @@ pub use crate::datastructures::vecmatrix::VecMatrix;
 
 /// nalgebra DMatrix based implementation
 pub use crate::datastructures::nalgebra::NAMatrix;
+
+/// helper for MPI calls to send both some accumulated cost and the rank who calculated it
+#[derive(Default, Clone, Copy, Equivalence)]
+#[cfg(feature = "mpi")]
+pub struct MPICostRank(pub f64, pub i32);
```
Comment (johann-cm) [inline src/datastructures/mod.rs]: I suggest using the mpi `Rank` type, to stay as platform-independent as possible:
```suggestion
pub struct MPICostRank(pub f64, pub mpi::topology::Rank);
```

File: src/solvers/approximate/nearest_neighbour.rs [hunk 2]
```diff
@@ -105,7 +103,28 @@ pub fn nearest_neighbour<const MODE: usize>(graph_matrix: &NAMatrix) -> Solution
...
+/// chunk locally and indenpendent.
+///
+/// Once the minimum of that local chunk has been found, the MPI code starts.
+/// First, we do an ALL_REDUCE on the cost so that every node knows the best solution.
+/// In the same REDUCE, we also give every node the winning node.
+///
+/// After that, the winner process knows that it won. It then proceedes to broadcast the whole path
+/// to everyone. This is an optimization as we just lazily send whole arrays through the network.
+///
+/// Now every node knows the global minimum and can successfully return it.
```
Comment (johann-cm) [inline src/solvers/approximate/nearest_neighbour.rs:127]: Very detailed and understandable doc-comment :+1:

**Comment by johann-cm** [comment]
All in all very nice code :+1:

**Pull request merged**